one-pager-maker/src/repository/shared/utils.ts

```diff
@@ -0,0 +1,104 @@
+import {
+    endAt,
+    endBefore,
+    limit,
+    orderBy,
+    QueryConstraint,
+    startAfter,
+    startAt, Timestamp,
+    where
+} from "firebase/firestore";
+
+export type OrderByDirection = 'asc' | 'desc';
+export type OrderBy<T> = {
+    field: keyof T,
+    direction: OrderByDirection
+}
+
+export type WhereFilter<T, K extends keyof T> = {
+    field: K,
+    op: WhereFilterOp,
+    value: T[K],
+}
+type WhereFilterOp =
+    '<'
+    | '<='
+    | '=='
+    | '!='
+    | '>='
+    | '>'
+    | 'array-contains'
+    | 'in'
+    | 'array-contains-any'
+    | 'not-in';
+
+export type QueryParams<T> = {
+    orderBy?: OrderBy<T>,
+    startAt?: unknown
+    startAfter?: unknown
+    endAt?: unknown,
+    endBefore?: unknown,
+    limit?: number,
+    where?: WhereFilter<T, keyof T> | WhereFilter<T, keyof T>[]
+}
+
+/**
+ * QueryParams を QueryConstraint[] に変換します
+ * @param query
+ */
+export function buildQueryConstraints<T>(query: QueryParams<T>) {
+    const constraints: QueryConstraint[] = [];
+
+    if (query.orderBy && typeof query.orderBy.field === 'string') {
+        constraints.push(orderBy(query.orderBy.field, query.orderBy.direction))
+    }
+    if (query.startAt) constraints.push(startAt(query.startAt))
+    if (query.startAfter) constraints.push(startAfter(query.startAfter))
+    if (query.endAt) constraints.push(endAt(query.endAt))
+    if (query.endBefore) constraints.push(endBefore(query.endBefore))
+    if (query.limit) constraints.push(limit(query.limit))
+
+    const addWhereConstraint = (whereFilter: WhereFilter<T, keyof T>) => {
+        if (typeof whereFilter.field === 'string')
```

Comment:
`query.orderBy.field` and  `whereFilter.field` are checked type of 'string' .  
Would there be any issues if we fixed the types of  `query.orderBy.field` and  `whereFilter.field` to string instead of using generics in the definition?

Reply:
I defined `query.orderBy.field` as `keyof T` to use it like `a[query.orderBy.field]`. I couldn't change it to `stringP.